tools/src/main/scala/akka/persistence/jdbc/tools/migration/LegacyJournalDataMigrator.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2014 - 2019 Dennis Vriend <https://github.com/dnvriend>
+ * Copyright (C) 2019 - 2021 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.jdbc.tools.migration
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.persistence.{ AtomicWrite, Persistence, PersistentRepr }
+import akka.persistence.journal.EventAdapter
+import akka.stream.scaladsl.Source
+import com.typesafe.config.Config
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.util.Try
+
+/**
+ * This will help migrate the legacy journal data onto the new journal schema with the
+ * appropriate serialization
+ *
+ * @param config the application config
+ * @param system the actor system
+ */
+final case class LegacyJournalDataMigrator(config: Config)(implicit system: ActorSystem) extends DataMigrator(config) {
+
+  private val eventAdapters = Persistence(system).adaptersFor("", config)
+
+  private def adaptEvents(repr: PersistentRepr): Seq[PersistentRepr] = {
+    val adapter: EventAdapter = eventAdapters.get(repr.payload.getClass)
+    adapter.fromJournal(repr.payload, repr.manifest).events.map(repr.withPayload)
+  }
```

Comment:
This is neat. 

This will allow users to eventually migrate their events to some new format.

There one thing that we should watch-out though. If an adapter produces more events then it takes (1 event split in many), that will change the ordering in the new table. 

That's not an issue in itself unless users are tracking the offset using Lagom or Akka Projections. 

That's all good, but we need to carefully inform users that their adapters should emit 1:1 in that case or they should reset their offset table and consume from scratch. 

There is one more thing on the adapters approach, it that we can re-tag the events.

Reply:
Interesting.